File: elements-of-programming-interviews/python/prime_sieve.py
from typing import List

from test_framework import generic_test


# Given n, return all primes up to and including n.
def generate_primes(n: int) -> List[int]:
    table = [False for i in range(n+1)]
    primes = []
    for i in range(2, n+1):
        if table[i] == True:
            continue
        primes.append(i)
        for j in range(0, n+1, i):
            table[j] = True
    return primes


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('prime_sieve.py', 'prime_sieve.tsv',
                                       generate_primes))

File: elements-of-programming-interviews/2023/cpp/prime_sieve.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;
// Brute force method. Given n return all primes up to and including n.
vector<int> GeneratePrimesBruteForce(int n) {
  int counter = 0;
  if (n == 0 || n == 1) {
    return {};
  }
  vector<int> primes{2};
  for (int i = 2; i <= n; i++) {
    int j = 0;
    for (; j < primes.size(); j++) {
      if (!(i % primes[j])) {
        break;
      }
    }
    if (j == primes.size()) {
      primes.push_back(i);
    }
  }
  return primes;
}

vector<int> GeneratePrimes(int n) {
  // Sieve of Eratosthenes
  vector<int> primes;
  vector<bool> table(n + 1, false);
  for (size_t i = 2; i <= n; i++) {
    if (table[i] == true) {
      continue;
    }
    primes.push_back(i);
    for (size_t j = 0; j <= n; j += i) {
      table[j] = true;
    }
  }
  return primes;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"n"};
  return GenericTestMain(args, "prime_sieve.cc", "prime_sieve.tsv",
                         &GeneratePrimes, DefaultComparator{}, param_names);
}
